crates/bevy_pbr/src/fog.rs

```diff
@@ -0,0 +1,488 @@
+        ///
+        /// Each component can be thought of as an independent per `R`/`G`/`B` channel `density` factor from
+        /// [`FogFalloff::Exponential`]: Multiplier applied to the world distance (within the fog
+        /// falloff calculation) for that specific channel.
+        ///
+        /// **Note:**
+        /// This value is not a `Color`, since it affects the channels exponentially in a non-intuitive way.
+        /// For artistic control, use the [`FogFalloff::from_visibility_colors()`] convenience method.
+        extinction: Vec3,
+
+        /// Controls how much light is added due to light scattering from the sun through the atmosphere.
+        ///
+        /// Each component can be thought of as an independent per `R`/`G`/`B` channel `density` factor from
+        /// [`FogFalloff::Exponential`]: A multiplier applied to the world distance (within the fog
+        /// falloff calculation) for that specific channel.
+        ///
+        /// **Note:**
+        /// This value is not a `Color`, since it affects the channels exponentially in a non-intuitive way.
+        /// For artistic control, use the [`FogFalloff::from_visibility_colors()`] convenience method.
+        inscattering: Vec3,
+    },
+}
+
+impl FogFalloff {
+    /// Creates a [`FogFalloff::Exponential`] value from the given visibility distance in world units,
+    /// using the revised Koschmieder contrast threshold, [`FogFalloff::REVISED_KOSCHMIEDER_CONTRAST_THRESHOLD`].
+    pub fn from_visibility(visibility: f32) -> FogFalloff {
+        FogFalloff::from_visibility_contrast(
+            visibility,
+            FogFalloff::REVISED_KOSCHMIEDER_CONTRAST_THRESHOLD,
+        )
+    }
+
+    /// Creates a [`FogFalloff::Exponential`] value from the given visibility distance in world units,
+    /// and a given contrast threshold in the range of `0.0` to `1.0`.
+    pub fn from_visibility_contrast(visibility: f32, contrast_threshold: f32) -> FogFalloff {
+        FogFalloff::Exponential {
+            density: FogFalloff::koschmieder(visibility, contrast_threshold),
+        }
+    }
+
+    /// Creates a [`FogFalloff::ExponentialSquared`] value from the given visibility distance in world units,
+    /// using the revised Koschmieder contrast threshold, [`FogFalloff::REVISED_KOSCHMIEDER_CONTRAST_THRESHOLD`].
+    pub fn from_visibility_squared(visibility: f32) -> FogFalloff {
+        FogFalloff::from_visibility_contrast(
+            visibility,
+            FogFalloff::REVISED_KOSCHMIEDER_CONTRAST_THRESHOLD,
+        )
+    }
+
+    /// Creates a [`FogFalloff::ExponentialSquared`] value from the given visibility distance in world units,
+    /// and a given contrast threshold in the range of `0.0` to `1.0`.
+    pub fn from_visibility_contrast_squared(
+        visibility: f32,
+        contrast_threshold: f32,
+    ) -> FogFalloff {
+        FogFalloff::Exponential {
+            density: FogFalloff::koschmieder(visibility, contrast_threshold).sqrt(),
```

Comment:
```suggestion
            density: FogFalloff::koschmieder(visibility, contrast_threshold) / 2.0,
```

Reply:
actually hang on, this isn't right either... i assumed exp squared was using `exp(distance*density)^2` but it's actually using `exp((distance*density)^2)`.

so then the density to recover `contrast_threshold` at distance `visibility` must be `sqrt(linear density / z)` i think?
```suggestion
            density: (FogFalloff::koschmieder(visibility, contrast_threshold) / visibility).sqrt(),
```

Reply:
I've double-checked the maths and indeed `(FogFalloff::koschmieder(visibility, contrast_threshold) / visibility).sqrt()` in the last suggestion from @robtfm in the previous message above is correct.

Reply:
Good catch btw!

Reply:
Oh, really nice catch! I must have botched the math along the way when isolating the density and didn't realize it